### PR TITLE
Add multi role support for API and model

### PR DESCRIPTION
**Pull request by ANeumann82**
Add new fields for multi role support

Summary:
Add role field to app, pod and instance
Validate role field
Set default role on instances if no role is explicitly stated

JIRA issues:
MARATHON-8637
MARATHON-8638

### CODE REVIEW

**Comment by jeschkies** [comment]
Over one thousand lines is quite a bit. We usually try to stay under 500 lines of changes*. Shall we pair later to see if we can split this up?

* I know, I violated that rules myself quite a bit 😉

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala [hunk 1]
```diff
@@ -89,9 +104,18 @@ class AppsResource @Inject() (
...
       implicit val identity = await(authenticatedAsync(req))
 
-      val rawApp = Raml.fromRaml(Json.parse(body).as[raml.App].normalize)
+      val rawApp = Raml.fromRaml(Json.parse(body).as[raml.App].normalize(validateAndNormalizeApp))
       val now = clock.now()
-      val app = validateOrThrow(rawApp).copy(versionInfo = VersionInfo.OnlyVersion(now))
+
+      // This is not really thread safe, another thread may intercept us and change the enforceRole flag, so we need
+      // to revalidate this inside the the groupManager later
+      val roleEnforcement = getEnforcedRoleForService(rawApp.id)
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: What do you think about using a `Validator[Role]` or something similar instead? It could still be constructed dending on `group.enforceRole` but would still be in line with the validation we use.

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala [hunk 2]
```diff
@@ -89,9 +104,18 @@ class AppsResource @Inject() (
...
+      val appDefinitionValidator = AppDefinition.validAppDefinition(config.availableFeatures, roleEnforcement)(pluginManager)
+
+      // TODO AN: This should be somewhere else... Normalization maybe?
+      val appWithRole = if (rawApp.role.isDefined) rawApp else rawApp.copy(role = Some(roleEnforcement.defaultRole))
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: I don't really understand this line. This results from role being optional I guess. See my comment above.

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 3]
```diff
@@ -486,13 +496,26 @@ object AppDefinition extends GeneralPurposeCombinators {
...
-    enabledFeatures: Set[String])(implicit pluginManager: PluginManager): Validator[AppDefinition] =
+    enabledFeatures: Set[String], roleEnforcement: RoleEnforcement)(implicit pluginManager: PluginManager): Validator[AppDefinition] =
     validator[AppDefinition] { app =>
       app.id is valid and PathId.absolutePathValidator and PathId.nonEmptyPath
       app.dependencies is every(PathId.pathIdValidator)
-
+      app is validWithRoleEnforcement(roleEnforcement)
     } and validBasicAppDefinition(enabledFeatures) and pluginValidators
 
+  private def validWithRoleEnforcement(roleEnforcement: RoleEnforcement): Validator[AppDefinition] = validator[AppDefinition] { app =>
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: I wonder if we should not call this instead of `getRoleEnforcement` in `AppsResource`:
```
  private def validWithRoleEnforcement(enforceRole: Boolean, validRoles: Set[String]): Validator[AppDefinition] = validator[AppDefinition] { app =>
  ...
}
```

What do you think?

File: src/main/scala/mesosphere/marathon/state/Group.scala [hunk 4]
```diff
@@ -143,12 +143,12 @@ object Group extends StrictLogging {
   def defaultDependencies: Set[PathId] = Set.empty
   def defaultVersion: Timestamp = Timestamp.now()
 
-  def validGroup(base: PathId, enabledFeatures: Set[String]): Validator[Group] =
+  def validGroup(base: PathId, config: MarathonConf): Validator[Group] =
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/Group.scala]: What is the motivation for this change?

File: src/main/scala/mesosphere/marathon/state/Instance.scala [hunk 5]
```diff
@@ -20,7 +20,8 @@ case class Instance(
     state: InstanceState,
     tasksMap: Map[Task.Id, Task],
     runSpecVersion: Timestamp,
-    reservation: Option[Reservation]) extends MarathonState[Protos.Json, Instance] {
+    reservation: Option[Reservation],
+    role: String) extends MarathonState[Protos.Json, Instance] {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/Instance.scala]: What do you think about moving the changes for instance out into a different pull request? They seem orthogonal to me.

File: src/main/scala/mesosphere/marathon/state/Group.scala (hunk 4 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/state/Group.scala]: We now need the config.mesosRole inside the GroupValidator, not only the config.enabledFeatures

File: docs/docs/rest-api/public/api/v2/types/app.raml [hunk 6]
```diff
@@ -326,6 +326,12 @@ types:
         type: boolean
         description: |
           Describes if (pseudo) TTY sould be allocated for the process of this container.
+      role?:
+        type: string
+        minLength: 1
+        description: |
+          The role to use. If not specified, uses the role of the enclosing group, or the default role.
+          At the moment, only the default role or the group role are allowed
```
[the remark this answers is not in the document]
Comment (ANeumann82) [reply docs/docs/rest-api/public/api/v2/types/app.raml]: Hmmm, in the end, there are only two possible values, ```--mesos_role``` or top-level-group-role. 
For legacy/unmigratable applications, i.e. ones that have a custom acceptedResourceRoles set, the value should be the one from the acceptedResourceRoles, as far as I can understand it...

File: src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala [hunk 7]
```diff
@@ -73,31 +73,32 @@ object AppHelpers {
     partialUpdate: Boolean,
     allowCreation: Boolean,
     now: Timestamp,
-    service: MarathonSchedulerService)(implicit
-    identity: Identity,
-    authorizer: Authorizer,
+    service: MarathonSchedulerService,
     appDefinitionValidator: Validator[AppDefinition],
-    appNormalization: Normalization[raml.App]): AppDefinition = {
+    appNormalization: Normalization[raml.App])(implicit
```
[the remark this answers is not in the document]
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala]: Because I'm now creating the validator for each invocation customized. And didn't think about that I can still have it implicit. Changed it back.

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala (hunk 2 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: Yeah. It basically sets the appropriate default based on the enclosing group's setting.

File: src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala]: Ah, ok. To be honest, I don't like it to be implict myself. However, this would be less distracting changes.

File: src/main/scala/mesosphere/marathon/state/RunSpec.scala [hunk 8]
```diff
@@ -24,6 +24,22 @@ case class BackoffStrategy(
...
+  * @param enforceRole If true, the role on the runSpec must match the specified role, additionally the
+  *                    acceptedResourceRole field can only contain one of ['*', &lt;role&gt;]
+  * @param validRoles List of valid roles
+  */
+case class RoleEnforcement(
+    enforceRole: Boolean = false,
+    validRoles: Set[String] = Set.empty) {
+
+  def defaultRole = validRoles.head
+}
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/RunSpec.scala]: Hm, this does not feel right. Also, a set is not sorted so `validRoles.head` seems to invoke undefined behavior for `defaultRole`.

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 9]
```diff
@@ -85,7 +85,9 @@ case class AppDefinition(
 
     override val killSelection: KillSelection = KillSelection.DefaultKillSelection,
 
-    tty: Option[Boolean] = AppDefinition.DefaultTTY) extends RunSpec
+    tty: Option[Boolean] = AppDefinition.DefaultTTY,
+
+    role: Option[String] = AppDefinition.DefaultRole) extends RunSpec
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: I don't think this should be optional.

Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: In that case, what do we use for DefaultRole? Or do we just not provide a default role and change every invocation to provide some role?

File: src/main/scala/mesosphere/marathon/state/RunSpec.scala (hunk 8 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/state/RunSpec.scala]: Yeah, it doesn't. Maybe it'd be better to have an explicit defaultRole here.

File: src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala]: Yeah, that's why i reverted it for now. Making this PR a little bit more readable is more important right now

File: src/main/scala/mesosphere/marathon/state/RunSpec.scala (hunk 8 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/state/RunSpec.scala]: Replaced it with an explicit defaultRole

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueDelegate.scala [hunk 10]
```diff
@@ -31,8 +31,10 @@ private[launchqueue] class LaunchQueueDelegate(
   override def notifyOfInstanceUpdate(update: InstanceChange): Future[Done] =
     askQueueActorFuture[InstanceChange, Done]("notifyOfInstanceUpdate")(update)
 
-  override def add(runSpec: RunSpec, count: Int): Future[Done] =
+  override def add(runSpec: RunSpec, count: Int): Future[Done] = {
+    require(runSpec.role.isDefined, "Adding a runSpec to launch queue requires a valid role in runSpec")
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueDelegate.scala]: Why make `runSpec.role` optional? I think it should be always defined so why not express it in the type system?

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala [hunk 11]
```diff
@@ -79,6 +78,22 @@ class AppsResource @Inject() (
     }
   }
 
+  private def getEnforcedRoleForService(servicePathId: PathId): RoleEnforcement = {
```
[the remark this answers is not in the document]
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: Done

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala (hunk 2 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: This is now in the normalization

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala (hunk 9 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: Non-Optional now

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueDelegate.scala (hunk 10 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueDelegate.scala]: Non-Optional now

File: src/main/scala/mesosphere/marathon/util/RoleUtils.scala [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+  *
+  * @param enforceRole If true, the role on the runSpec must match the specified role, additionally the
+  *                    acceptedResourceRole field can only contain one of ['*', &lt;role&gt;]
+  * @param validRoles List of valid roles
+  * @param defaultRole The default role to use if no role is specified on the service. Defaults to '*'
+  */
+case class RoleSettings(
+    enforceRole: Boolean = false,
+    validRoles: Set[String] = Set.empty,
+    defaultRole: String = ResourceRole.Unreserved) {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/util/RoleUtils.scala]:
```suggestion
    defaultRole: String) {
```

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala [hunk 13]
```diff
@@ -349,7 +372,13 @@ class AppsResource @Inject() (
       case (_, false) => CompleteReplacement
     }
 
+    val roleSettings = RoleUtils.getRoleSettingsForService(config, appId, groupManager.rootGroup())
+    val normalizationConfig = AppNormalization.Configuration(config, roleSettings)
+    implicit val appDefinitionValidator: Validator[AppDefinition] = AppDefinition.validAppDefinition(config.availableFeatures, roleSettings)(pluginManager)
+    implicit val validateAndNormalizeApp: Normalization[raml.App] = appNormalization(normalizationConfig)(AppNormalization.withCanonizedIds())
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: There's lots or repetition. What do you think about moving this into a helper method?

File: src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala [hunk 14]
```diff
@@ -48,19 +48,16 @@ class PodsResource @Inject() (
...
-    PodsValidation.podValidator(
-      config.availableFeatures,
-      scheduler.mesosMasterVersion().getOrElse(SemanticVersion(0, 0, 0)), config.defaultNetworkName.toOption)
 
   // If we change/add/upgrade the notion of a Pod and can't do it purely in the internal model,
   // update the json first
-  private implicit val normalizer = PodNormalization.apply(PodNormalization.Configuration(
-    config.defaultNetworkName.toOption))
+  //  private implicit val normalizer = PodNormalization.apply(PodNormalization.Configuration(
+  //    config.defaultNetworkName.toOption))
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala]: Is this a left over?

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 15]
```diff
@@ -135,22 +135,23 @@ trait PodsValidation extends GeneralPurposeCombinators {
...
 
-  val artifactValidator = validator[Artifact] { artifact =>
+  private val artifactValidator = validator[Artifact] { artifact =>
     artifact.uri.length is between(1, 1024)
     artifact.destPath.map(_.length).getOrElse(1) is between(1, 1024)
   }
 
-  val lifeCycleValidator = validator[Lifecycle] { lc =>
+  // TODO: This is unused???
+  private val lifeCycleValidator = validator[Lifecycle] { lc =>
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: Just try it out 🙂

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 16]
```diff
@@ -250,11 +253,30 @@ trait PodsValidation extends GeneralPurposeCombinators {
...
     pod should haveUnreachableDisabledForResidentPods
-    pod should haveValidAcceptedResourceRoles
+    pod should haveValidAcceptedResourceRoles(roleEnforcement.validRoles)
+    pod is validWithRoleEnforcement(roleEnforcement)
+  }
+
+  private def validWithRoleEnforcement(roleEnforcement: RoleSettings): Validator[Pod] = validator[Pod] { pod =>
+    if (roleEnforcement.enforceRole) {
+      pod.role must notEmpty
+      pod.role.orNull as "role" is in(roleEnforcement.validRoles) // We need to use orNull here, as accord validator-and does not short-circuit
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]:
```suggestion
      pod.role as "role" is optional(in(roleEnforcement.validRoles))
```

File: src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerImpl.scala [hunk 17]
```diff
@@ -129,13 +129,18 @@ class GroupManagerImpl(
           case Left(left) =>
             Left(left)
           case Right(changed) =>
+            // Set default roles
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerImpl.scala]: Is this a todo? Or should it be part of #6959?

File: src/main/scala/mesosphere/marathon/raml/AppConversion.scala [hunk 18]
```diff
@@ -137,6 +138,11 @@ trait AppConversion extends DefaultConversions with CheckConversion with Constra
 
     val versionInfo = state.VersionInfo.OnlyVersion(app.version.map(Timestamp(_)).getOrElse(Timestamp.now()))
 
+    val role = app.role.orNull
+    if (role == null) {
+      throw new IllegalArgumentException("Failed to deserialize raml.App, no role provided")
+    }
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/raml/AppConversion.scala]: This should be more Scala like
val role = app.role.getOrElse(throw new IllegalArgumentException("Failed to convert raml.App, no role provided. This is a bug in AppNormalization."))

File: src/main/scala/mesosphere/marathon/raml/PodConversion.scala [hunk 19]
```diff
@@ -43,6 +43,11 @@ trait PodConversion extends NetworkConversion with ConstraintConversion with Con
 
     val executorResources: ExecutorResources = podd.executorResources.getOrElse(PodDefinition.DefaultExecutorResources.toRaml)
 
+    val role = podd.role.orNull
+    if (role == null) {
+      throw new IllegalArgumentException("Deserializing raml.Pod failed, as no role was provided.")
+    }
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/raml/PodConversion.scala]: See above

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 20]
```diff
@@ -459,13 +468,15 @@ object AppDefinition extends GeneralPurposeCombinators {
 
   val DefaultTTY: Option[Boolean] = None
 
+  val DefaultRole: Option[String] = None
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: Is this still used?

File: src/main/scala/mesosphere/marathon/state/RootGroup.scala [hunk 21]
```diff
@@ -408,12 +410,27 @@ object RootGroup {
     RootGroup(group.apps, group.pods, group.groupsById, group.dependencies, group.version)
   }
 
-  def rootGroupValidator(enabledFeatures: Set[String]): Validator[RootGroup] = {
-    noCyclicDependencies and
-      Group.validGroup(PathId.empty, enabledFeatures) and
-      ExternalVolumes.validRootGroup()
+  def rootGroupValidator(config: MarathonConf): Validator[RootGroup] = validator[RootGroup] { rootGroup =>
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/state/RootGroup.scala]: What do you think about renaming this to `validRootGroup`?

File: src/main/scala/mesosphere/marathon/util/RoleUtils.scala [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+object RoleSettings {
+  def forTest = RoleSettings(validRoles = Set(ResourceRole.Unreserved), defaultRole = ResourceRole.Unreserved)
+}
+
+object RoleUtils extends StrictLogging {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/util/RoleUtils.scala]:
```suggestion
object RoleSettings extends StrictLogging {
```

File: src/main/scala/mesosphere/marathon/util/RoleUtils.scala [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+  def getRoleSettingsForService(config: MarathonConf, servicePathId: PathId, rootGroup: RootGroup): RoleSettings = {
+    val defaultRole = config.mesosRole()
+
+    // We have a service in the root group, no enforced role here
+    if (servicePathId.parent.isRoot) return RoleSettings(validRoles = Set(defaultRole), defaultRole = defaultRole)
+
+    val topLevelGroupPath = servicePathId.rootPath
+
+    rootGroup.group(topLevelGroupPath).map(group => {
+      if (group.enforceRole.getOrElse(false)) {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/util/RoleUtils.scala]: Related to #6959 and should be `conf.defaultEnforceGroupRole()`.

File: src/main/scala/mesosphere/marathon/util/RoleUtils.scala [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+        RoleSettings(enforceRole = true, validRoles = Set(group.id.root), defaultRole = group.id.root)
+      } else {
+        RoleSettings(validRoles = Set(defaultRole, group.id.root), defaultRole = defaultRole)
+      }
+    }).getOrElse(RoleSettings(validRoles = Set(defaultRole), defaultRole = defaultRole))
+  }
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/util/RoleUtils.scala]:
```scala
val topLevelGroup = rootGroup.group(servicePathId.rootPath)
val enforceRole = topLevelGroup.enforceRole.get
if (enforceRole) {
  RoleSettings(enforceRole = enforceRole, validRoles = Set(topLevelGroup.id), defaultRole = topLevelGroup.id)
} else {
  RoleSettings(enforceRole = enforceRole, validRoles = Set(confg.mesosRole(), topLevelGroup.id), defaultRole = config.mesosRole())
}

}
```

File: src/main/scala/mesosphere/marathon/util/RoleUtils.scala (hunk 12 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/util/RoleUtils.scala]: Done

File: src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala (hunk 14 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/api/v2/PodsResource.scala]: Yes, removed

File: src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerImpl.scala (hunk 17 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerImpl.scala]: It's a leftover from the previous approach. Removed it.

File: src/main/scala/mesosphere/marathon/raml/AppConversion.scala (hunk 18 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/raml/AppConversion.scala]: Ah, right. Changed.

File: src/main/scala/mesosphere/marathon/raml/PodConversion.scala (hunk 19 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/raml/PodConversion.scala]: Fixed

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala (hunk 20 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: Nope, removed it

File: src/main/scala/mesosphere/marathon/util/RoleUtils.scala (hunk 22 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/util/RoleUtils.scala]: Good idea, changed it

File: src/main/scala/mesosphere/marathon/util/RoleUtils.scala (hunk 23 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/util/RoleUtils.scala]: I'll leave this as is for now, as there's no defaultEnforceGroupRole in this branch

File: src/main/scala/mesosphere/marathon/state/RootGroup.scala (hunk 21 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/state/RootGroup.scala]: Done

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala (hunk 15 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: Unused, removed it

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala (hunk 16 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: Seems to work ;)

File: src/main/scala/mesosphere/marathon/util/RoleUtils.scala (hunk 24 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/util/RoleUtils.scala]: It's actually even more complicated. I've adjusted the code a bit, we certainly need tim to have a look at it...

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala (hunk 13 above)
Comment (ANeumann82) [reply src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: Yeah, moved the validator/normalizer-creation into two helper methods, one for apps one for appUpdates
